docs/proposals/shuffle-sharding-on-the-read-path.md

```diff
@@ -0,0 +1,350 @@
+
+The problem to solve is: how can a querier efficiently find which ingesters have data for a given tenant?  Each option must consider the changing of the set of ingesters and the changing of each tenant’s subring size.
+
+### Option #1: Ingesters expose list of tenants
+
+A possible solution could be keeping in the querier an in-memory data structure to map each ingester to the list of tenants for which it has some data. This data structure would be constructed at querier startup, and then periodically updated, interpolating two information:
+
+1. The current state of the ring
+2. The list of tenants directly exposed by each ingester (via a dedicated gRPC call)
+
+#### Scenario: new querier starts up
+
+When a querier starts up and before getting ready:
+
+1. It scans all ingesters (discovered via the ring) and fetches the list of tenants for which each ingester has some data
+2. For each found tenant (unique list of tenant IDs across all ingesters responses), the querier looks at the current state of the ring and adds to the map the list of ingesters currently assigned to the tenant shard, even if they don’t hold any data yet (because may start receiving series shortly)
+
+Then the querier watches the ingester ring and rebuilds the in-memory map whenever the ring topology changes.
+
+#### Scenario: querier receives a query for an unknown tenant
+
+A new tenant starts remote writing to the cluster. The querier doesn’t know it in its in-memory map, so it adds the tenant on the fly to the map just looking at the current state of the ring.
+
+#### Scenario: ingester scale up / down
+
+When a new ingester is added / removed to / from the ring, the ring topology changes and queriers will update the in-memory map.
+
+#### Scenario: per-tenant shard size increases
+
+Queriers periodically (every 1m) reload the limits config file. When a tenant shard size change is detected, the querier updates the in-memory map for the affected tenant.
+
+**Issue:** some time series data may be missing in queries up to 1m.
+
+#### Edge case: queriers notice the ring topology change before distributors
+
+Consider the following scenario:
+
+1. Tenant A shard is composed by ingesters 1,2,3,4,5,6
+2. Tenant A is remote writing 1 single series and gets replicated to ingester 1,2,3
+3. The ring topology changes and tenant A shard is ingesters 1,2,3,7,8,9
+4. Querier notices the ring topology change and updates the in-memory map. Given tenant A series were only on ingester 1,2,3, the querier maps tenant A to ingester 1,2,3 (because of what received from ingesters via gRPC) and 7,8,9 (because of the current state of the ring)
+5. Distributor hasn’t updated the ring state yet
+6. Tenant A remote writes 1 **new** series, which get replicated to 4,5,6
+7. Distributor updates the ring state
+8. **Race condition:** querier will not know that ingesters 4,5,6 contains tenant A data until the next sync
+
+### Option #2: use only the information contained in the ring.
```

Comment:
Since we agreed to move forward with this, I would not call it "option" but "proposal" and then I would rename options 1 and 2 as "alternatives we considered".